File: PA103.java
//Leetcode 116. Populating Next Right Pointers in Each Node
//Time Complexity: O(n) , n is no of nodes in the tree
//Space Complexity: O(1) , As stack space is not be counted as extra space(mentioned in que) 
class Solution {
    public Node connect(Node root) {
        if(root==null) return root;        
        helper( root);
        return root;
        
    }
    public void helper(Node root) {
        if(root==null || root.left==null || root.right==null) return;      
        root.left.next= root.right;
        if(root.next!=null) root.right.next= root.next.left;
        helper(root.left);        
        helper(root.right);
        
            
    }
}


==================================
//BFS solution
//Time Complexity: O(n) , n is no of nodes in the tree
//Space Complexity:O(n), space used for queue

class Solution {
    public Node connect(Node root) {
        if(root==null) return root;
        
        Queue<Node> q = new LinkedList<>();
        q.offer(root);
        
        while(!q.isEmpty()){
           int s= q.size();
           Node prev=null;
           while(s>0){
               Node n= q.poll();  
               if(n.left!=null) q.offer(n.left);
               if(n.right!=null) q.offer(n.right);
               if(prev!=null) prev.next=n;
               prev=n;
               s--;
            }
        }
        return root;
        
    }
}

File: PA103.py
#Time Complexity - O(h)
#Space Complexity - O(h)
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=0, left=None, right=None, next=None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
from collections import deque
class Solution(object):
    def connect(self, root):
        """
        :type root: Node
        :rtype: Node
        """
        if not root:
            return root
        tmp = deque()
        tmp.append(root)
        t1 = root
        while tmp:
            l1 = len(tmp)
            prev = None
            for i in range(l1):
                curr = tmp.popleft()
                curr.next = None
                if not prev:
                    prev = curr
                    prev.next = None
                else:
                    prev.next = curr
                    prev = prev.next
                    prev.next=None
                
                if curr.left:
                    tmp.append(curr.left)
                if curr.right:
                    tmp.append(curr.right)
        return(root)